<a href="https://colab.research.google.com/github/SujitbMusale/Loan-Predctions---Logistic-Regression/blob/main/Loan_Prediction_Classification_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [68]:
import pandas as pd
import datetime as dt
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.metrics import precision_recall_fscore_support


import warnings
warnings.filterwarnings("ignore")

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
df = pd.read_csv('/content/drive/MyDrive/almabetter/csv files/LoanStats Dataset.csv',encoding ='latin1')

In [4]:
df.head()

Loan ID  Amount Requested  Amount Funded By Investors Interest Rate  \
0   87023            7500.0                      7500.0        13.75%   
1   90376            5000.0                      5000.0         7.43%   
2   90395            5000.0                      5000.0         8.07%   
3   90665            8500.0                      8500.0        10.28%   
4   92187            2500.0                      2500.0         8.07%   

  Loan Length Application Date Application Expiration Date Issued Date  \
0   36 months       2007-06-07                  2007-06-21  2007-06-27   
1   36 months       2007-06-12                  2007-06-26  2007-07-02   
2   36 months       2007-06-12                  2007-06-26  2007-07-02   
3   36 months       2007-06-13                  2007-06-27  2007-07-03   
4   36 months       2007-06-15                  2007-06-29  2007-07-05   

  CREDIT Grade                 Loan Title  ... Inquiries in the Last 6 Months  \
0           E2         Consolidation Loan  ...                            0.0   
1           A2                   JAL Loan  ...                            0.0   
2           A4     MBA Loan Consolidation  ...                            0.0   
3           C1  Retiring credit card debt  ...                            1.0   
4           A4           Home Improvement  ...                            0.0   

  Accounts Now Delinquent  Delinquent Amount Delinquencies (Last 2 yrs)  \
0                     0.0                0.0                        1.0   
1                     0.0                0.0                        0.0   
2                     0.0                0.0                        0.0   
3                     0.0                0.0                        1.0   
4                     0.0                0.0                        0.0   

   Months Since Last Delinquency Public Records On File  \
0                           11.0                    0.0   
1                            0.0                    0.0   
2                            0.0                    0.0   
3                            5.0                    0.0   
4                            0.0                    0.0   

   Months Since Last Record  Education  Employment Length  Code  
0                       0.0        NaN           < 1 year   NaN  
1                       0.0        NaN           < 1 year   NaN  
2                       0.0        NaN           < 1 year   NaN  
3                       0.0        NaN            3 years   NaN  
4                       0.0        NaN            4 years   NaN  

[5 rows x 41 columns]

In [5]:
df.shape

(36396, 41)

In [6]:
df.iloc[1]

Loan ID                                         90376
Amount Requested                               5000.0
Amount Funded By Investors                     5000.0
Interest Rate                                   7.43%
Loan Length                                 36 months
Application Date                           2007-06-12
Application Expiration Date                2007-06-26
Issued Date                                2007-07-02
CREDIT Grade                                       A2
Loan Title                                   JAL Loan
Loan Purpose                                    other
Loan Description                                  NaN
Monthly PAYMENT                                155.38
Status                                     Fully Paid
Total Amount Funded                            5000.0
Debt-To-Income Ratio                            3.72%
Remaining Principal Funded by Investors           0.0
Payments To Date (Funded by investors)        5174.16
Remaining Principal         

In [7]:
# lets check the columns present in the dataset
# number of the columns of the dataset

print('features  are: ',df.columns)
print('\n')
print('The number of the features are: ',len(df.columns))

features  are:  Index(['Loan ID', 'Amount Requested', 'Amount Funded By Investors',
       'Interest Rate', 'Loan Length', 'Application Date',
       'Application Expiration Date', 'Issued Date', 'CREDIT Grade',
       'Loan Title', 'Loan Purpose', 'Loan Description', 'Monthly PAYMENT',
       'Status', 'Total Amount Funded', 'Debt-To-Income Ratio',
       'Remaining Principal Funded by Investors',
       'Payments To Date (Funded by investors)', 'Remaining Principal ',
       ' Payments To Date', 'Screen Name', 'City', 'State', 'Home Ownership',
       'Monthly Income', 'FICO Range', 'Earliest CREDIT Line',
       'Open CREDIT Lines', 'Total CREDIT Lines', 'Revolving CREDIT Balance',
       'Revolving Line Utilization', 'Inquiries in the Last 6 Months',
       'Accounts Now Delinquent', 'Delinquent Amount',
       'Delinquencies (Last 2 yrs)', 'Months Since Last Delinquency',
       'Public Records On File', 'Months Since Last Record', 'Education',
       'Employment Length', 'Code'],

so total 41 features are present in the dataset


In [8]:
# lets check the null values present in the data set
print(df.isnull().sum())

Loan ID                                        0
Amount Requested                               1
Amount Funded By Investors                     1
Interest Rate                                  1
Loan Length                                    1
Application Date                               1
Application Expiration Date                    1
Issued Date                                  177
CREDIT Grade                                   1
Loan Title                                    14
Loan Purpose                                   1
Loan Description                           10544
Monthly PAYMENT                                1
Status                                         1
Total Amount Funded                            1
Debt-To-Income Ratio                           1
Remaining Principal Funded by Investors        1
Payments To Date (Funded by investors)         1
Remaining Principal                            1
 Payments To Date                              1
Screen Name         

we can see that there are many variables are having null values

In [9]:
# lets check the unique values present in the various independent and dependent varaible
# lest refer the unique value list as follows
lst=[]
for i in df.columns:
  lst.append([i,len(df[i].unique())])
unique_feature_list=pd.DataFrame(lst,columns=['Feature','No of unique element'])
print(unique_feature_list.sort_values(by='No of unique element'))

                                    Feature  No of unique element
32                  Accounts Now Delinquent                     3
4                               Loan Length                     3
33                        Delinquent Amount                     4
23                           Home Ownership                     6
25                               FICO Range                     7
36                   Public Records On File                     7
39                        Employment Length                    12
34               Delinquencies (Last 2 yrs)                    13
10                             Loan Purpose                    15
13                                   Status                    19
31           Inquiries in the Last 6 Months                    29
8                              CREDIT Grade                    36
27                        Open CREDIT Lines                    45
22                                    State                    51
28        

In [10]:
#lets check the data type present in the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36396 entries, 0 to 36395
Data columns (total 41 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Loan ID                                  36396 non-null  object 
 1   Amount Requested                         36395 non-null  float64
 2   Amount Funded By Investors               36395 non-null  float64
 3   Interest Rate                            36395 non-null  object 
 4   Loan Length                              36395 non-null  object 
 5   Application Date                         36395 non-null  object 
 6   Application Expiration Date              36395 non-null  object 
 7   Issued Date                              36219 non-null  object 
 8   CREDIT Grade                             36395 non-null  object 
 9   Loan Title                               36382 non-null  object 
 10  Loan Purpose                             36395

In [11]:
# lest check the unique values in Status feature
df['Status'].unique()

array(['Fully Paid', 'Charged Off', 'Late (31-120 days)',
       'Performing Payment Plan', 'Default', 'In Grace Period', 'Current',
       'Late (16-30 days)', 'Issued', 'In Review', nan,
       'Does not meet the current credit policy  Status: Fully Paid',
       'Does not meet the current credit policy  Status: Performing Payment Plan',
       'Does not meet the current credit policy  Status: Charged Off',
       'Does not meet the current credit policy  Status: Late (31-120 days)',
       'Does not meet the current credit policy  Status: In Grace Period',
       'Does not meet the current credit policy  Status: Late (16-30 days)',
       'Does not meet the current credit policy  Status: Current',
       'Does not meet the current credit policy  Status: Default'],
      dtype=object)

In [12]:
#remove all other records except the  'Fully paid', 'Charged off', 'Late (31-120 days), 'Default', 'In Grace Period' and  'Current' records
data_with_loanstatus_sliced = df[(df['Status']=="Fully Paid") | (df['Status']=="Charged Off") |
                                   (df['Status']=='Does not meet the current credit policy  Status: Fully Paid') | (df['Status']=='Does not meet the current credit policy  Status: Charged Off') | (df['Status']=="Late (31-120 days)") | 
                                   (df['Status']=="Default") | (df['Status']=="In Grace Period") | (df['Status']=="Current")| (df['Status']=="Issued")| (df['Status']=="Does not meet the current credit policy  Status: Current")|
                                   (df['Status']=="Does not meet the current credit policy  Status: Late (31-120 days)")| (df['Status']=="Does not meet the current credit policy  Status: In Grace Period")|(df['Status']=="Does not meet the current credit policy  Status: Default")]

In [13]:
# lets change class value with the integer 
val =  {"Fully Paid":0, "Charged Off":5, 'Does not meet the current credit policy  Status: Fully Paid':0, 
        'Does not meet the current credit policy  Status: Charged Off':5, 'Late (31-120 days)':2, 'Default': 4, 
        'In Grace Period':3, 'Current':1,'Issued':6, 'Does not meet the current credit policy  Status: Current':1, 
        'Does not meet the current credit policy  Status: Late (31-120 days)':2,
        'Does not meet the current credit policy  Status: In Grace Period':3,
        'Does not meet the current credit policy  Status: Default':4} 
df=data_with_loanstatus_sliced.replace({"Status":val})

In [14]:
# convert the "Earliest CREDIT Line" attribute to  Python Date time object.
df["Earliest CREDIT Line"]=pd.to_datetime(df['Earliest CREDIT Line'])

In [15]:
# treat the null value present in the "Earliest CREDIT Line".
dd=int(df['Earliest CREDIT Line'].dt.day.median())
mm=int(df['Earliest CREDIT Line'].dt.month.median())
yy=int(df['Earliest CREDIT Line'].dt.year.median())

In [16]:
d=dt.datetime(yy,mm,dd)
date =d.date()

In [17]:
# replacing null value in 'Earliest CREDIT Line' with median date
df['Earliest CREDIT Line']=df['Earliest CREDIT Line'].fillna(date)

In [18]:
# check the null value in the 'Earliest CREDIT Line' 
df['Earliest CREDIT Line'].isnull().any()

False

In [19]:
# convert 'Application Date' to date-time attribute
df['Application Date']=pd.to_datetime(df['Application Date'])
dd=int(df['Application Date'].dt.day.median())
mm=int(df['Application Date'].dt.month.median())
yy=int(df['Application Date'].dt.year.median())
d = dt.datetime(yy,mm,dd)
date = d.date()
df['Application Date'] = df['Application Date'].fillna(date)

In [20]:
# convert 'Application Expiration Date' to date-time attribute

df['Application Expiration Date'] = pd.to_datetime(df['Application Expiration Date'])
dd=int(df['Application Expiration Date'].dt.day.median())
mm=int(df['Application Expiration Date'].dt.month.median())
yy=int(df['Application Expiration Date'].dt.year.median())
d = dt.datetime(yy,mm,dd)
date = d.date()
df['Application Expiration Date'] = df['Application Expiration Date'].fillna(date)

In [21]:
# converting 'Issue Date' to date-time attribute
df['Issued Date'] = pd.to_datetime(df['Issued Date'])
dd=int(df['Issued Date'].dt.day.median())
mm=int(df['Issued Date'].dt.month.median())
yy=int(df['Issued Date'].dt.year.median())
d = dt.datetime(yy,mm,dd)
date = d.date()
df['Issued Date'] = df['Issued Date'].fillna(date)

In [22]:
# converting the string value to numeric value
# intrest rate
df['Interest Rate'] =df['Interest Rate'].replace({'%':''},regex=True)
df['Interest Rate']= df['Interest Rate'].apply(lambda x : float(x))

In [23]:
# just to check
print(df['Interest Rate'][0:5])
print('\n')
print(df['Interest Rate'].dtype)

0    13.75
1     7.43
2     8.07
3    10.28
4     8.07
Name: Interest Rate, dtype: float64


float64


In [24]:
# convert 'Debt-To-Income Ratio' to numerical datatype
df['Debt-To-Income Ratio'] = df['Debt-To-Income Ratio'].replace({'%':''}, regex = True)
df['Debt-To-Income Ratio'] = df['Debt-To-Income Ratio'].apply(lambda x:float(x))

In [25]:
# convert 'Revolving Line Utilization' to numerical datatype
df['Revolving Line Utilization'] = df['Revolving Line Utilization'].replace({'%':''}, regex = True)
df['Revolving Line Utilization'] = df['Revolving Line Utilization'].apply(lambda x:float(x))

In [26]:
# lets work with the attributes with redentent information and extract numerical data and transform to numeric value
df['Loan Length'] = df['Loan Length'].replace({' months':''}, regex = True)
df['Loan Length'] = df['Loan Length'].apply(lambda x:int(x))

In [27]:
# lets work with categorical variable
# lets remove the space in 'CREDIT Grade' to 'CREDIT_Grade'
df.rename(columns = {'CREDIT Grade':'CREDIT_Grade'}, inplace = True)
df.rename(columns = {'Home Ownership':'Home_Ownership'}, inplace = True)
df.rename(columns = {'FICO Range':'FICO_Range'}, inplace = True)


In [28]:
# let's check the unique values present in the above variable 
print('unique value in CREDIT_Grade: ')
print(df['CREDIT_Grade'].unique())
print('\n')
print('unique value in Home_Ownership: ')
print(df['Home_Ownership'].unique())
print('\n')
print('unique value in FICO_Range: ')
print(df['FICO_Range'].unique())

unique value in CREDIT_Grade: 
['E2' 'A2' 'A4' 'C1' 'B1' 'C4' 'B5' 'B3' 'A5' 'A3' 'B2' 'C2' 'F4' 'C5'
 'E1' 'D3' 'B4' 'D1' 'F1' 'D4' 'C3' 'D2' 'F2' 'A1' 'E4' 'F3' 'E5' 'D5'
 'E3' 'G5' 'F5' 'G4' 'G3' 'G2' 'G1']


unique value in Home_Ownership: 
['OWN' 'MORTGAGE' 'RENT' 'NONE' 'ANY']


unique value in FICO_Range: 
['660-678' '780+' '714-749' '679-713' '750-779' nan '640-659']


In [29]:
# lets replace the above test value by numerical values
# replcae the value as "640-659":6,
                        #"660-678":5,
                        #"679-713":4,
                        #"714-749":3,
                        #"750-779":2,
                        #"780+":1
df["FICO_Range"] = df["FICO_Range"].map({"640-659":6,"660-678":5,"679-713":4,"714-749":3,"750-779":2,"780+":1})


In [30]:
# also there are some null value lets treat them
print(df['FICO_Range'].value_counts())
print('\n')
print('null value presents are: ')
print(df['FICO_Range'].isnull().sum())

4.0    12508
3.0    10250
5.0     5660
2.0     4898
1.0     2064
6.0      472
Name: FICO_Range, dtype: int64


null value presents are: 
17


In [31]:
# replace null value with 3
df['FICO_Range']=df['FICO_Range'].fillna(3)

In [32]:
# lets study the Education Attributes
print(len(df['Education'].unique()))
print('\n')
print(df['Education'].unique()[0:10])
print('\n')
print('null value present tin the attribute')
print(df['Education'].isnull().sum())

3622


[nan 'Youngstown State University' 'University of Connecticut'
 'Rochester Institute of Technology (RIT)' 'Penn State-University Park'
 'Missouri State University'
 'Columbia University in the City of New York,Harvard University'
 'Aquinas College' 'Western Washington University'
 'Georgia Institute of Technology']


null value present tin the attribute
30085


In [33]:
# lets replace null value with zero and all education details with 1
df['Education']=df['Education'].fillna(0)
df['Education']=df['Education'].apply(lambda x : 0 if x == 0 else 1 )

In [34]:
df['Education'].unique()

array([0, 1])

In [35]:
# lets work on Employment Length attribute
# first rename the attribute
df.rename(columns={'Employment Length':'Employment_Length'},inplace=True)


In [36]:
# lets check the unique value present in the Employment_Length
print(df['Employment_Length'].unique())
print('\n')
print('null value presents are: ')
print(df['Employment_Length'].isnull().sum())

['< 1 year' '3 years' '4 years' '5 years' '1 year' '2 years' '8 years'
 '7 years' '6 years' '9 years' '10+ years' nan]


null value presents are: 
875


In [37]:
# fill null value with zero
df['Employment_Length'] = df['Employment_Length'].fillna(0)

# lets convert output text variable to numeric variable.
# '< 1 year':1, 
# '1 year':2, 
# '2 years':3, 
# '3 years':4, 
# '4 years':5, 
# '5 years':6, 
# '7 years':8, 
# '6 years':7, 
# '8 years':9, 
# '9 years':10, 
# '10+ years':11

df['Employment_Length'] = df['Employment_Length'].map({'< 1 year':1, '3 years':4, 
                                                       '4 years':5, '5 years':6, '1 year':2, 
                                                       '2 years':3, '8 years':9, '7 years':8, 
                                                       '6 years':7, '9 years':10, '10+ years':11})



In [38]:
df['Employment_Length'][0:5]

0    1.0
1    1.0
2    1.0
3    4.0
4    5.0
Name: Employment_Length, dtype: float64

In [39]:
# lets work with 'Loan Purpose' attribute
# rename the column name
df.rename(columns={'Loan Purpose':'Loan_Purpose'},inplace=True)

In [40]:
# lets check the unique value present in the 'Loan_Purpose' attribute
print(len(df['Loan_Purpose'].unique()))
print('\n')
print(df['Loan_Purpose'].value_counts())

14


debt_consolidation    16270
credit_card            4376
other                  3855
home_improvement       2808
major_purchase         2113
small_business         1699
car                    1445
wedding                 886
medical                 647
moving                  557
educational             417
house                   357
vacation                343
renewable_energy         96
Name: Loan_Purpose, dtype: int64


In [41]:
# now for the remaining fill na with mean of that attributes column
df=df.fillna(df.mean())

Attribute selection for Modeling

In [42]:
# we will not consider below attribute in the modeling
dcolumns = ['Loan ID', 'Amount Funded By Investors','Loan Title','Loan_Purpose',
            'Loan Description','Total Amount Funded','Remaining Principal Funded by Investors',
            'Payments To Date (Funded by investors)',
            'Screen Name','City','Code' ,'Application Date', 'Application Expiration Date', 'Issued Date','Earliest CREDIT Line',
           'Inquiries in the Last 6 Months','Accounts Now Delinquent','Delinquent Amount','Delinquencies (Last 2 yrs)',
            'Months Since Last Delinquency','Public Records On File','Months Since Last Record',
           'CREDIT_Grade','State','Home_Ownership','FICO_Range','Education','Employment_Length']
# drop the above columns from the modeling
df=df.drop(dcolumns,axis=1)

In [43]:
df.head()

Amount Requested  Interest Rate  Loan Length  Monthly PAYMENT  Status  \
0            7500.0          13.75           36           255.43       0   
1            5000.0           7.43           36           155.38       0   
2            5000.0           8.07           36           156.84       0   
3            8500.0          10.28           36           275.38       0   
4            2500.0           8.07           36            78.42       0   

   Debt-To-Income Ratio  Remaining Principal    Payments To Date  \
0                 14.29                   0.0            9194.98   
1                  3.72                   0.0            5174.16   
2                  2.30                   0.0            5272.08   
3                  6.40                   0.0            9912.49   
4                 11.33                   0.0            2822.72   

   Monthly Income  Open CREDIT Lines  Total CREDIT Lines  \
0         1833.33                7.0                 8.0   
1        16666.67               17.0                26.0   
2         8333.33               11.0                20.0   
3         1500.00                6.0                 9.0   
4         9166.67               13.0                40.0   

   Revolving CREDIT Balance  Revolving Line Utilization  
0                    4175.0                        51.5  
1                   85607.0                         0.7  
2                    9698.0                        19.4  
3                    8847.0                        26.9  
4                    7274.0                        13.1

* we can see from above dataset the dataset is clean with required attributes

In [44]:
# separate Independent and Depedent variabele
X=df.drop(['Status'],axis=1)
Y=df['Status']

In [45]:
# check the data shape
print(X.shape,Y.shape)

(35869, 12) (35869,)


* So we are having 35869 records which are good with 12 attributes in data

In [46]:
# lets normalize the dataset for further process
scaler = MinMaxScaler()
X=scaler.fit_transform(X)

In [57]:
# lets split the data to train test data
X_train,X_test,y_train,y_test = train_test_split(X,Y,test_size=0.2,random_state=42)

In [58]:
# take Logistic regression attribute
lr = LogisticRegression(multi_class='multinomial', random_state=1)


In [59]:
lr.fit(X_train,y_train)

LogisticRegression(multi_class='multinomial', random_state=1)

In [60]:
y_pred = lr.predict(X_test)

In [61]:
y_test[0:10]

16415    1
9063     1
15364    1
25956    1
34467    0
10465    0
13169    1
26206    1
26290    1
14730    2
Name: Status, dtype: int64

In [62]:
y_pred[0:10]

array([1, 1, 1, 1, 0, 0, 1, 1, 1, 1])

In [63]:
print(confusion_matrix(y_test,y_pred))

[[1135    0    0    0    0    0    0]
 [  99 4938    0    0    0    0    7]
 [   0   87    0    0    0    0    0]
 [   6   40    0    0    0    0    0]
 [   1    7    0    0    0    0    0]
 [   6  311    0    0    0    1    1]
 [   0  486    0    0    0    0   49]]


In [69]:
print(precision_recall_fscore_support(y_test,y_pred,average='weighted'))

(0.8441373637615257, 0.8534987454697519, 0.7996778000218187, None)


from above tupple we can say that the precision is 84.41% , recall is 85.34% and f1 score is 79.96% 

In [ ]:
# now we will perform hyper parameter tunning to improve the all performance matrix
